libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -156,6 +157,57 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 	}
 }
 
+void ContractLevelChecker::checkAbstractModifiers(ContractDefinition const& _contract)
```

Comment:
This is similar to `checkAbstractFunctions(...)`, with the following differences.

1. Doesn't care about stateVariables. TBH, I still haven't figured out exactly why `contract->stateVariables()` are checked in the function. If the same reason applies to modifiers, then some reworking is required.
2. Skips checks about contract (redundant; will be performed in checkAbstractFunctions.)

Reply:
To explain this: Public state variables automatically generate "getter" functions, so in the following contract:
```
contract C {
  uint public x;
}
```
The public state variable ``x`` implicitly generates a function that pretty much behaves like the following:
```
function x() external returns (uint256) { return x; }
```

If the public state variable is even declared ``public override``, then it can actually implement an abstract function of the same name and signature. That's why those are registered in ``checkAbstractFunctions``. Nothing like that can happen for modifiers, so it should be fine. I'm merely wondering, whether this whole process can be simplified and we can avoid code duplication - I'll have a look into that.

Reply:
Can you have a look at ``libsolidity/analysis/OverrideChecker.h`` and the ``OverrideProxy`` defined there?
I'm wondering, whether we can extract the ``OverrideProxy`` from there into its own file and reuse it here.

The ``OverrideProxy`` is an artificial wrapper for everything that behaves "function-like", in particular functions, public state variables and modifers.

Since we can assume that no modifier will have the same name as a public state variable or a function (this is already guaranteed by the name-and-type-resolver phase), we should be able to use that wrapper here as well to check both functions, public state variables and modifiers in one go without code duplication.

I'm not entirely sure this will work without issues, though - so don't spend too much time trying - if it's not straightforward, I could live with the duplication here.

Reply:
@ekpyron The repetitive code would be the lambda registerModifier, and the `std::map modifiers` can possibly be combined to `std::map functions` with the help of `OverrideProxy.` So we can combine the two functions `checkAbstractModifiers` and `checkAbstractFunctions` into one. This would probably make the resulting function more complex, so not sure if I should proceed with it.

On a related note, we can actually get rid of ` std::vector<ModifierDefinition const*> unimplementedModifiers` from `ContractDefinitionAnnotation` and keep it local to the function. What do you think about this?

Reply:
I think if we combine the two functions using override proxies, the only new thing that will be left is the additional case
```
		for (ModifierDefinition const* modifier: contract->functionModifiers())
			registerModifier(*modifier, TypeProvider::modifier(*modifier));
```

We need the unimplemented modifiers in the annotation, so that it can be properly exported in the json AST - which is by the way still missing in this PR ;-) - search for usages of the ``unimplementedFunctions`` annotation.

We could, however, change it to a single annotation ``unimplementedDeclarations`` for both functions and modifiers, if we want to merge.